daemon/cmd/devices.go

```diff
@@ -0,0 +1,296 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2016-2021 Authors of Cilium
+
+// This module implements Cilium's network device detection.
+
+package cmd
+
+import (
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/datapath/linux/probes"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/mac"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+)
+
+var (
+	excludedDevicePrefixes = []string{
+		"cilium_",
+		"lo",
+		"lxc",
+		"cni",
+		"docker",
+	}
+)
+
+type DeviceManager struct {
+	mu      lock.Mutex
+	devices map[string]bool
+}
+
+func NewDeviceManager() *DeviceManager {
+	return &DeviceManager{
+		devices: make(map[string]bool),
+	}
+}
+
+func (dm *DeviceManager) Close() {
+}
+
+// Detect tries to detect devices which are to be used for:
+// NodePort BPF, direct routing in NodePort BPF, IPsec and Bandwidth Manager.
+//
+// The devices are detected by looking at all the configured global unicast
+// routes in the system.
+func (dm *DeviceManager) Detect() error {
+	dm.mu.Lock()
+	defer dm.mu.Unlock()
+	dm.devices = make(map[string]bool)
+
+	if err := expandDevices(); err != nil {
+		log.WithError(err).Warnf("Failed to expand device wildcards")
+	}
+
+	l3DevOK := false
+	if option.Config.EnableNodePort {
+		// Probe whether fast redirect is supported for L3 devices. This will
+		// invoke bpftool and requires root privileges.
+		l3DevOK = supportL3Dev()
+	}
+
+	detectDevices := len(option.Config.Devices) == 0 &&
+		(option.Config.EnableNodePort || option.Config.EnableHostFirewall || option.Config.EnableBandwidthManager)
+
+	if detectDevices {
+		// Detect the devices from the system routing table by finding the devices
+		// which have global unicast routes.
+		family := netlink.FAMILY_ALL
+		if option.Config.EnableIPv4 && !option.Config.EnableIPv6 {
+			family = netlink.FAMILY_V4
+		} else if !option.Config.EnableIPv4 && option.Config.EnableIPv6 {
+			family = netlink.FAMILY_V6
+		}
+
+		routes, err := netlink.RouteList(nil, family)
+		if err != nil {
+			return fmt.Errorf("cannot retrieve routes for device detection: %w", err)
+		}
+		dm.updateDevicesFromRoutes(l3DevOK, routes)
+	} else {
+		for _, dev := range option.Config.Devices {
+			dm.devices[dev] = true
+		}
+	}
+
+	detectDirectRoutingDev := option.Config.EnableNodePort
+	if option.Config.DirectRoutingDevice != "" {
+		dm.devices[option.Config.DirectRoutingDevice] = true
+		detectDirectRoutingDev = false
+	}
+
+	detectIPv6MCastDev := option.Config.EnableIPv6NDP
+	if option.Config.IPv6MCastDevice != "" {
+		dm.devices[option.Config.IPv6MCastDevice] = true
+		detectIPv6MCastDev = false
+	}
+
+	k8sNodeDev := ""
+	k8sNodeLink, err := findK8SNodeIPLink()
+	if err == nil {
```

Comment:
No since k8s may not actually be always available in all use-cases. Though may be it should be here?

Reply:
Do you mean there's one error case that's expected? If that's the case, we should ignore that specific error and log the rest I think.

Reply:
Well, it can fail in two ways: either the Node IP is not set (k8s not enabled, or something else horrible wrong), or the device can not be found that has the Node IP. Right now we only care about it failing to find the device if we need that information for the direct routing device or mcast device detection. 

I guess I see the point of failing more eagerly here. I'll think about what would be more exact thing to do here.

Reply:
If we're just worried about environment where K8s is not installed, we could simply check `k8s.IsEnabled()` before looking up `k8sNodeLink`.

Reply:
Added the check and returning a better error now.